### PR TITLE
Add static analysis

### PR DESCRIPTION
**Pull request by sigiesec**
I have added static analysis builds experimentally:
- Under Linux on travis-ci
-- clang-tidy/clang-static-analyzer (currently 6.0, but should be easy to update to 7.0)
-- sonar-scanner
- Under Windows on appveyor
-- Visual Studio 2017 Static Analyzer
-- Visual Studio 2017 C++ Core Guidelines Check

I would like to get some feedback on this? Does it make sense? I think it could be a good idea to run these as scheduled builds, e.g. on a weekly basis.

The sonar-scanner results are published to sonarcloud.io (https://sonarcloud.io/dashboard?id=libzmq-clang, but I am not sure if this is publicly accessible, if not and you have/create an account on sonarcloud.io, I can give you access).

The other results are currently only visible in the build log, e.g. here:
* https://travis-ci.org/sigiesec/libzmq/jobs/380206165
* https://ci.appveyor.com/project/sigiesec/libzmq/build/build-628/job/5bi0inexkp6rsn7b

Originally, I thought I could publish these to sonarcloud.io as well. There is ta plugin for SonarQube, sonar-cxx, which supports this, but it is not available on sonarcloud.io, unfortunately.

I have already disabled some checks, which do not make sense on the current codebase, but this requires some more fine-tuning.

Many things are style issues, and primarily affect readability/maintainability.

The Visual Studio Static Analyzer shows several warnings, which sound like false positives:
* C6001: Using uninitialized memory
* C6387: 'data_' could be '0'
* C6385: Reading invalid data
* C6386: Buffer overrun
If the false positive rate is really that large, these checks should probably be disabled.

However, the clang-static-analyzer shows a small number of similar errors, which sound actually like bugs to me (at least it is not obvious that these situations cannot happen):
* decoder_allocators.cpp:121:5: warning: Use of memory after it is freed [clang-analyzer-unix.Malloc] (I already fixed this)
* generic_mtrie_impl.hpp:144:13: warning: Branch condition evaluates to a garbage value [clang-analyzer-core.uninitialized.Branch]
* stream_engine.cpp:815:9: warning: Use of memory after it is freed [clang-analyzer-cplusplus.NewDelete]
* ypipe.hpp:165:9: warning: Called C++ object pointer is null [clang-analyzer-core.CallAndMessage]
* proxy.cpp:376:14: warning: Called C++ object pointer is null [clang-analyzer-core.CallAndMessage]

Before this could be merged, at least the following would need to be done:
- [x] Set up sonarqube for zeromq org
- [x] Fine-tune active checks
- [x] Make execution of static analysis builds conditional (on specifically named branches, or on daily/weekly builds on zeromq/master)

### CODE REVIEW

**Comment by bluca** [comment]
Yeah I think this is a good idea - it's good data to have, and some of the warnings could be simple starter tasks.

File: src/zmq.cpp [hunk 1]
```diff
@@ -1176,16 +1176,18 @@ void *zmq_poller_new (void)
...
+            return 0;
+        }
     }
-
-    delete ((zmq::socket_poller_t *) poller);
-    *poller_p_ = NULL;
-    return 0;
+    errno = EFAULT;
+    return -1;
+}
```
Comment (bluca) [inline src/zmq.cpp]: copypasta? extra }

Comment (sigiesec) [reply src/zmq.cpp]: Yeah, sorry I was in a kind of haste ;)

**Comment by sigiesec** [comment]
@bluca Could you add a daily or weekly cron job to build master as described here? https://docs.travis-ci.com/user/cron-jobs/#Adding-Cron-Jobs

**Comment by bluca** [comment]
Sure, done

**Comment by sigiesec** [comment]
@bluca Thanks. Can you also set a SONARQUBE_TOKEN encrypted environment variable in the travis-ci configuration? I can send you the token via some other channel.

**Comment by bluca** [comment]
Sure, you can find my PGP key on the usual keyserver, the fingerprint is ```A9EA 9081 724F FAE0 484C  35A1 A81C EA22 BC8C 7E2E```
Alternatively I have the OTR plugin via IRC

**Pull request merged**